modules/sockets/sockets.cpp

```diff
@@ -15,233 +11,370 @@
-#define SOCKET_UDP 2
-
-// And global Variables:
-
-// native socket_open(_hostname[], _port, _protocol = SOCKET_TCP, &_error);
-static cell AMX_NATIVE_CALL socket_open(AMX *amx, cell *params)  /* 2 param */
-{ 
-    unsigned int port = params[2];
-    int len;
-    char* hostname = MF_GetAmxString(amx,params[1],0,&len); // Get the hostname from AMX
-	cell *err = MF_GetAmxAddr(amx, params[4]);
-    if(len == 0) { // just to prevent to work with a nonset hostname
-        *err = 2;  // server unknown
-        return -1;
-    }
-    *err = 0; // params[4] is error backchannel
-    struct sockaddr_in server;
-    struct hostent *host_info;
-    unsigned long addr;
-    int sock=-1;
-    int contr;
-    // Create a Socket
-    sock = socket(AF_INET, params[3]==SOCKET_TCP?SOCK_STREAM:SOCK_DGRAM, 0);
-    if (sock < 0) {
-        // Error, couldn't create a socket, so set an error and return.
-        *err = 1;
-        return -1;
-    }
-    
-    // Clear the server structure (set everything to 0)
-    memset( &server, 0, sizeof (server));
-    // Test the hostname, and resolve if needed
-    if ((addr = inet_addr(hostname)) != INADDR_NONE) {
-        // seems like hostname is a numeric ip, so put it into the structure
-        memcpy( (char *)&server.sin_addr, &addr, sizeof(addr));
-    }
-    else {
-        // hostname is a domain, so resolve it to an ip
-        host_info = gethostbyname(hostname);
-        if (host_info == NULL) {
-            // an error occured, the hostname is unknown
-            *err = 2;  // server unknown
-            return -1;
-        }
-        // If not, put it in the Server structure
-        memcpy( (char *)&server.sin_addr, host_info->h_addr, host_info->h_length);
-    }
-    // Set the type of the Socket
-    server.sin_family = AF_INET;
-    // Change the port to network byte order, and put it into the structure
-    server.sin_port = htons(port);
-    
-    // Not, let's try to open a connection to the server
-    contr = connect(sock, (struct sockaddr*)&server, sizeof( server));
-    if (contr < 0) {
-            // If an error occured cancel
-            *err = 3;  //error while connecting
-            return -1;
-    }
-    // Everything went well, so return the socket
-    return sock;
+	freeaddrinfo(server_info);
+
+	if(sockfd == -1 || server == nullptr)
+		return -1;
+
+	return sockfd;
 }
 
-// native socket_close(_socket);
-static cell AMX_NATIVE_CALL socket_close(AMX *amx, cell *params)  /* 2 param */
+int set_nonblocking(int sockfd)
+{
+#ifdef _WIN32
+	unsigned long flags = 1;
+
+	if(ioctlsocket(sockfd, FIONBIO, &flags) == 0)
+		return 0;
+	else
+		return errno;
+#else
+	int flags = -1;
+
+	if((flags = fcntl(sockfd, F_GETFL, 0)) == -1)
+		return errno;
+
+	if(fcntl(sockfd, F_SETFL, flags | O_NONBLOCK) == -1)
+		return errno;
+
+	return 0;
+#endif
+}
+
+// native socket_open_nb(_hostname[], _port, _protocol = SOCKET_TCP, &_error);
+static cell AMX_NATIVE_CALL socket_open_nb(AMX *amx, cell *params)
```

Comment:
Maybe be can use something like:
```SourcePawn
socket_open(_hostname[], _port, _protocol = SOCKET_TCP, &_error, _libc_errors = DISABLE_LIBC_ERRORS, _socket_type = BLOCKING);
```

It can't be done the way you said because 'connect' can block, so we can't call socket_open first and then set that socket a non-blocking. The socket should be set as non-blocking right away as it is created,